core/drivers/crypto/ele/ele.c

```diff
@@ -359,33 +313,65 @@ static TEE_Result imx_ele_session_close(uint32_t session_handle)
 	return imx_ele_call(&msg);
 }
 
-int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+static TEE_Result imx_ele_get_device_info(struct get_info_rsp *rsp)
 {
-	uint32_t session_handle = 0;
-	/*
-	 * The die ID must be cached because some board configuration prevents
-	 * the MU to be used by OPTEE at runtime.
-	 */
-	static struct session_get_device_info_rsp rsp;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct imx_ele_buf output = {};
+	struct {
+		uint32_t addr_msb;
+		uint32_t addr_lsb;
+		uint16_t size;
+	} __packed cmd = {};
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_BASELINE,
+		.header.size = SIZE_MSG_32(cmd),
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_GET_INFO,
+	};
 
-	if (rsp.rsp_code)
+	if (!rsp)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = imx_ele_buf_alloc(&output, NULL, sizeof(*rsp));
+	if (res)
 		goto out;
 
-	if (imx_ele_session_open(&session_handle))
-		goto err;
+	cmd.addr_msb = output.paddr_msb;
+	cmd.addr_lsb = output.paddr_lsb;
+	cmd.size = sizeof(*rsp);
 
-	if (imx_ele_session_get_device_info(session_handle, &rsp))
-		goto err;
+	memcpy(msg.data.u8, &cmd, sizeof(cmd));
+
+	res = imx_ele_call(&msg);
+	if (res)
+		goto out;
+
+	res = imx_ele_buf_copy(&output, (uint8_t *)rsp, sizeof(*rsp));
+out:
+	imx_ele_buf_free(&output);
+
+	return res;
+}
+
+int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+{
+	static uint32_t uid[4];
+	static bool is_fetched;
+	struct get_info_rsp rsp = {};
 
-	if (imx_ele_session_close(session_handle))
+	assert(buffer && len);
+
+	if (is_fetched)
+		goto out;
+
+	if (imx_ele_get_device_info(&rsp))
 		goto err;
 
+	memcpy(uid, rsp.uid, MIN(sizeof(rsp.uid), len));
+
+	is_fetched = true;
 out:
-	/*
-	 * In the device info array return by the ELE, the words 2, 3, 4 and 5
-	 * are the device UID.
-	 */
-	memcpy(buffer, rsp.chip_uid, MIN(sizeof(rsp.chip_uid), len));
+	memcpy(buffer, uid, MIN(sizeof(uid), len));
```

Comment:
Am I right to assume that the "die id" will remain unchanged, it's only a new way of obtaining it?

Reply:
Yes, you are right.